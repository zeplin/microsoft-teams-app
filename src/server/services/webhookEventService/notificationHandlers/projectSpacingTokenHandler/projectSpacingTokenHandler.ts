import {
    WebhookEvent,
    ProjectSpacingTokenCreateEvent,
    ProjectSpacingTokenUpdateEvent
} from "../../../../adapters/zeplin/types";
import { NotificationHandler } from "../NotificationHandler";
import { SHORT_DELAY } from "../constants";
import { commonTeamsCard, MessageCard } from "../teamsCardTemplates";
import { md } from "../md";
import { getRandomEmoji } from "../getRandomEmoji";
import { getRedirectURLForZeplinApp, getWebAppURL } from "../zeplinURL";
import { getSpacingTokenUpdateMessage } from "../getStyleUpdateMessage";

type Event = ProjectSpacingTokenCreateEvent | ProjectSpacingTokenUpdateEvent;

class ProjectSpacingTokenHandler extends NotificationHandler<Event> {
    delay = SHORT_DELAY;
    private getText(events: Event[]): string {
        const [{
            payload: {
                action,
                context: {
                    project: {
                        name: projectName
                    }
                },
                resource: {
                    data: {
                        name: pivotSpacingTokenName
                    }
                }
            }
        }] = events;
        const actionText = action === "created" ? "added to" : "updated in";
        return events.length === 1
            ? md`**${pivotSpacingTokenName}** is ${actionText} _${projectName}_! ${getRandomEmoji()}`
            : md`**${events.length} spacing tokens** are ${actionText} _${projectName}_! ${getRandomEmoji()}`;
    }

    private getSectionText(events: Event[]): string {
        const [{
            payload: {
                context: {
                    project: {
                        platform: projectPlatform
                    }
                }
            }
        }] = events;

        return getSpacingTokenUpdateMessage(projectPlatform);
    }

    private getWebappURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        const pathname = `project/${projectId}/styleguide/spacing`;
        const searchParams = {
            sptid: events.map(event => event.payload.resource.id)
        };

        return getWebAppURL(pathname, searchParams);
    }

    private getMacAppURL(
        events: Event[]
    ): string {
        const [{
            payload: {
                context: {
                    project: {
                        id: projectId
                    }
                }
            }
        }] = events;
        const searchParams = {
            pid: projectId,
            sptids: events.map(event => event.payload.resource.id)
        };

        return getRedirectURLForZeplinApp("spacing", searchParams);
    }

    shouldHandleEvent(event: WebhookEvent): event is Event {
        return event.payload.action !== "deleted";
    }

    getTeamsMessage(
        events: Event[]
    ): MessageCard {
        return commonTeamsCard({
            text: this.getText(events),
            section: {
                text: this.getSectionText(events)
            },
            links: [{
                title: "Open in App",
                url: this.getMacAppURL(events)
            }, {
                title: "Open in Web",
                url: this.getWebappURL(events)
            }]
        });
    }
}

export const projectSpacingTokenHandler = new ProjectSpacingTokenHandler();
