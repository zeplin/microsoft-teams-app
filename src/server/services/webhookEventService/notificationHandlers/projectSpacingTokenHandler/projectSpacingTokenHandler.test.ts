import { projectSpacingTokenHandler } from "./projectSpacingTokenHandler";
import {
    ProjectPlatform,
    ProjectSpacingTokenCreateEvent,
    ProjectSpacingTokenUpdateEvent
} from "../../../../adapters/zeplin/types";

type GetDummyEventParams = {
    action?: string;
    spacingTokenId?: string;
    spacingTokenName?: string;
    projectPlatform?: ProjectPlatform;
}

function getDummyEvent({
    action = "created",
    spacingTokenId = "spacingTokenId",
    spacingTokenName = "spacingTokenName",
    projectPlatform = ProjectPlatform.WEB
}: GetDummyEventParams = {}): ProjectSpacingTokenCreateEvent | ProjectSpacingTokenUpdateEvent {
    return {
        payload: {
            action,
            context: {
                project: {
                    id: "projectId",
                    name: "projectName",
                    platform: projectPlatform
                }
            },
            resource: {
                id: spacingTokenId,
                data: {
                    id: spacingTokenId,
                    name: spacingTokenName
                }
            }
        }
    } as ProjectSpacingTokenCreateEvent | ProjectSpacingTokenUpdateEvent;
}

describe("projectSpacingTokenHandler", () => {
    describe("getTeamsMessage method", () => {
        beforeAll(() => {
            jest.spyOn(Math, "random").mockReturnValue(0);
        });

        describe("for created notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectSpacingTokenHandler.getTeamsMessage([getDummyEvent()])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectSpacingTokenHandler.getTeamsMessage([
                        getDummyEvent(),
                        getDummyEvent({
                            spacingTokenId: "anotherSpacingTokenId",
                            spacingTokenName: "anotherSpacingTokenName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each([
                ProjectPlatform.ANDROID,
                ProjectPlatform.IOS,
                ProjectPlatform.MAC_OS,
                ProjectPlatform.WEB
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectSpacingTokenHandler.getTeamsMessage([getDummyEvent({ projectPlatform })])
                    ).toMatchSnapshot();
                }
            );
        });

        describe("for updated notification", () => {
            it("should match snapshot when there is only 1 event", () => {
                expect(
                    projectSpacingTokenHandler.getTeamsMessage([getDummyEvent({ action: "updated" })])
                ).toMatchSnapshot();
            });

            it("should match snapshot when there are more than 1 events", () => {
                expect(
                    projectSpacingTokenHandler.getTeamsMessage([
                        getDummyEvent({
                            action: "updated"
                        }),
                        getDummyEvent({
                            action: "updated",
                            spacingTokenId: "anotherSpacingTokenId",
                            spacingTokenName: "anotherSpacingTokenName"
                        })
                    ])
                ).toMatchSnapshot();
            });

            it.each([
                ProjectPlatform.ANDROID,
                ProjectPlatform.IOS,
                ProjectPlatform.MAC_OS,
                ProjectPlatform.WEB
            ])("should match snapshot when project platform is %s",
                projectPlatform => {
                    expect(
                        projectSpacingTokenHandler.getTeamsMessage([
                            getDummyEvent({
                                action: "updated",
                                projectPlatform
                            })
                        ])
                    ).toMatchSnapshot();
                }
            );
        });
    });
});
