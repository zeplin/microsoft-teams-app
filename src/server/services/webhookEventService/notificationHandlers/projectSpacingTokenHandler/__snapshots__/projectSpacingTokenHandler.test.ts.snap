// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when project platform is android 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`spacing.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when project platform is ios 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when project platform is macos 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when project platform is web 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 spacing tokens** are added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%2CanotherSpacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&sptid=anotherSpacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for created notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is added to _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when project platform is android 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure \`spacing.xml\` is up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when project platform is ios 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when project platform is macos 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when project platform is web 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when there are more than 1 events 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**2 spacing tokens** are updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%2CanotherSpacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&sptid=anotherSpacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;

exports[`projectSpacingTokenHandler getTeamsMessage method for updated notification should match snapshot when there is only 1 event 1`] = `
{
  "@type": "MessageCard",
  "@context": "https://schema.org/extensions",
  "text": "**spacingTokenName** is updated in _projectName_! ðŸ—ž",
  "sections": [
    {
      "text": "Make sure your tokens are up to date!",
      "images": []
    }
  ],
  "potentialAction": [
    {
      "@type": "OpenUri",
      "name": "Open in App",
      "targets": [
        {
          "os": "default",
          "uri": "https://msteams-app.zeplin.io/api/app-redirect?uri=zpl%3A%2F%2Fspacing%3Fpid%3DprojectId%26sptids%3DspacingTokenId%26ref%3Dmicrosoft-teams"
        }
      ]
    },
    {
      "@type": "OpenUri",
      "name": "Open in Web",
      "targets": [
        {
          "os": "default",
          "uri": "https://app.zeplin.io/project/projectId/styleguide/spacing?sptid=spacingTokenId&ref=microsoft-teams"
        }
      ]
    }
  ]
}
`;
